pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceApp.java

```diff
@@ -0,0 +1,140 @@
+            streamManager.createStream(getConfig().getScope(), getConfig().getStream1Name(), streamConfig);
+            streamManager.createStream(getConfig().getScope(), getConfig().getStream2Name(), streamConfig);
+            // Create stream for the membership state synchronizer.
+            streamManager.createStream(
+                    getConfig().getScope(),
+                    getConfig().getMembershipSynchronizerStreamName(),
+                    StreamConfiguration.builder().scalingPolicy(ScalingPolicy.fixed(1)).build());
+        }
+
+        final ReaderGroupConfig readerGroupConfig = ReaderGroupConfig.builder()
+                .stream(Stream.of(getConfig().getScope(), getConfig().getStream1Name()))
+                .automaticCheckpointIntervalMillis(getConfig().getCheckpointPeriodMs())
+                .build();
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(getConfig().getScope(), clientConfig);
+        // Create the Reader Group (ignored if it already exists)
+        readerGroupManager.createReaderGroup(getConfig().getReaderGroup(), readerGroupConfig);
+        @Cleanup
+        final ReaderGroup readerGroup = readerGroupManager.getReaderGroup(getConfig().getReaderGroup());
+        @Cleanup
+        final EventStreamClientFactory eventStreamClientFactory = EventStreamClientFactory.withScope(getConfig().getScope(), clientConfig);
+        @Cleanup
+        final SynchronizerClientFactory synchronizerClientFactory = SynchronizerClientFactory.withScope(getConfig().getScope(), clientConfig);
+        // Create a Pravega stream writer that we will send our processed output to.
+        @Cleanup
+        final EventStreamWriter<SampleEvent> writer = eventStreamClientFactory.createEventWriter(
+                getConfig().getStream2Name(),
+                new JSONSerializer<>(new TypeToken<SampleEvent>(){}.getType()),
+                EventWriterConfig.builder().build());
+
+        final SampleEventProcessor processor = new SampleEventProcessor(
+                () -> ReaderGroupPruner.create(
+                        readerGroup,
+                        getConfig().getMembershipSynchronizerStreamName(),
+                        instanceId,
+                        synchronizerClientFactory,
+                        Executors.newScheduledThreadPool(1),
+                        getConfig().getHeartbeatIntervalMillis()),
+                () -> eventStreamClientFactory.<SampleEvent>createReader(
+                        instanceId,
+                        readerGroup.getGroupName(),
+                        new JSONSerializer<>(new TypeToken<SampleEvent>(){}.getType()),
+                        ReaderConfig.builder().build()),
+                1000,
+                instanceId,
+                writer);
+
+        processor.startAsync();
+
+        // Add shutdown hook for graceful shutdown.
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
```

Comment:
Shouldn't the processor do this itself if this is needed.?

Reply:
It gets messy to do this in the AtLeastOnceProcessor because then it would need to deal with removing the shutdown hook, but only doing that when the JVM is not actually shutting down. The point of this shutdown hook is to shutdown the whole application gracefully, which in this case happens to consist of only one AtLeastOnceProcessor service. But in general, an application may consist of many services that all need to be shutdown in a particular order. So it seems appropriate that AtLeastOnceApp should coordinate the shutdown.